Rewriting annex - interoperability with other programming models (#304)
openshmem-org/specification: content/interoperability.tex

    should also guarantee that the \openshmem and \ac{MPI} calls are not made concurrently
    from two distinct threads.
\end{itemize}

\subsection{Mapping Process Identification Numbers}
\label{subsec:interoperability:id}

Similar to the \ac{PE} number in \openshmem, \ac{MPI} defines rank as the
identification number of a process in a communicator. Both the \openshmem \ac{PE}
and the \ac{MPI} rank are unique integers assigned from zero to one less than the total
number of processes. In a hybrid program, the \openshmem
\ac{PE} number in \LibHandleRef{SHMEM\_TEAM\_WORLD}
and the \ac{MPI} rank in \VAR{MPI\_COMM\_WORLD} of a process can be equal.
This feature, however, may be provided by only some of the \openshmem and \ac{MPI}
implementations (e.g., if both environments share the same underlying process
manager) and is not portably guaranteed. A portable program should always
use the standard functions in each model, namely, \FUNC{shmem\_team\_my\_pe} in \openshmem
and \FUNC{MPI\_Comm\_rank} in \ac{MPI}, to query the process identification numbers
in each communication environment and manage the mapping of identifiers in the

> program when necessary.

Reviewer: Maybe it's implicit or maybe I'm missing something, but do we need to also specify that the size of `MPI_COMM_WORLD` is the same as `SHMEM_TEAM_WORLD`?  Otherwise, the example code below might not work correctly, right?  If we do need to specify the sizes are equal, do we have to also account for the usage of `MPI_Comm_Spawn`, which IIUC has a separate `MPI_COMM_WORLD` on the children ranks?

Author: Good question. You are right that there is no guarantee that size of `MPI_COMM_WORLD` is same as size of `SHMEM_TEAM_WORLD`. I think it depends on two factors:
1.  How the program initializes SHMEM and MPI (e.g., some processes call `MPI_Init` but do not call `shmem_init` )
2. How the implementations define accessible processes (e.g., although all launched processes call both SHMEM and MPI init, some of them might be visible in MPI but not visible in SHMEM).

For the example code, it forces each process to initialize both MPI and SHMEM, thus factor-(1) is confirmed. For factor-(2), I think only the processes who are accessible in both models can execute the code (e.g., any hidden process should not be able to execute the code after init calls). Thus, the size of each "WORLD" should be equal and the program works correctly.

For the program using `MPI_Comm_Spawn`, it should not use SHMEM to communicate with dynamically created processes. We have explicitly defined it in section "Dynamic Process Creation".

Please let me know if I miss any case.

Reviewer: Yes, I think you're correct - thanks!

      {example_code/hybrid_mpi_mapping_id.c}

\subsection{RMA Programming Models}
\label{subsec:interoperability:rma}

\openshmem and \ac{MPI} each define similar one-sided communication models;
however, a portable program should not assume interoperability between these
models.
For instance, \openshmem guarantees the atomicity only of concurrent \openshmem AMO operations
that operate on symmetric data with the same datatype. Access to the same symmetric
object with \ac{MPI} atomic operations, such as an \FUNC{MPI\_Fetch\_and\_op}, may
result in an undefined result. A hybrid program should avoid situations where \ac{MPI} and
\openshmem one-sided operations perform concurrent accesses to the same memory
location; otherwise, the behavior is undefined.

\subsection{Communication Progress}
\label{subsec:interoperability:progress}

\openshmem promises the progression of communication both with and without
\openshmem calls and requires the software progress mechanism in the implementation
(e.g., a progress thread) when the hardware does not provide asynchronous communication
capabilities. In \ac{MPI}, however, a weak progress semantics is applied. That is,

> an \ac{MPI} communication call is guaranteed only to complete in finite time. For

Author: Add reference to section 4.1.

\chapter{Interoperability with Other Programming Models}\label{sec:interoperability}

OpenSHMEM routines may be used in conjunction with the routines of other
communication libraries or parallel languages in the same program. This section
describes the interoperability with other programming models, including
clarification of undefined behaviors caused by mixed use of different models,
advice to \openshmem library users and developers that may improve the portability
and performance of hybrid programs, and definition of an OpenSHMEM
API that queries the interoperability features provided by an \openshmem library.

\section{MPI Interoperability}

\openshmem and MPI are two commonly used parallel programming models for
distributed-memory systems. The user can choose to utilize both models in the same program
to efficiently and easily support various communication patterns.

A vendor may implement the \openshmem and MPI libraries in different ways. For
instance, one may implement both \openshmem and MPI as standalone libraries,
each of which allocates and initializes fully isolated communication
resources.
Another common approach

> is to implement both \openshmem and MPI interfaces within the

Author: Fixed

a call to \FUNC{shmem\_init} or \FUNC{shmem\_init\_thread} and be finalized by
a call to \FUNC{shmem\_finalize}; the MPI environment of the program must be initialized
by a call to \FUNC{MPI\_Init} or \FUNC{MPI\_Init\_thread} and be finalized by a
call to \FUNC{MPI\_Finalize}.

\apiimpnotes{
Portable implementations of OpenSHMEM and MPI must ensure that the initialization
calls can be made in an arbitrary order within a program; the same rule also
applies to the finalization calls. A software runtime that utilizes a shared
communication resource for \openshmem and MPI communication may maintain an
internal reference counter in order to ensure that the shared resource is
initialized only once and thus no shared resource is released until the last
finalization call is made.
}

\subsection{Dynamic Process Creation}
\label{subsec:interoperability:mpmd}

MPI defines a dynamic process model that allows creation of processes after
an MPI application has started (e.g., by calling \FUNC{MPI\_Comm\_spawn}) and
connection to independent processes (e.g., through \FUNC{MPI\_Comm\_accept}
and \FUNC{MPI\_Comm\_connect})
and provides a mechanism to establish communication

> between the newly created processes and the existing MPI application (see

Author: Fixed

MPI standard version 3.1, Chapter 10).
Unlike MPI, \openshmem starts all processes at once and requires all PEs to
collectively allocate and initialize resources (e.g., symmetric heap) used by
the \openshmem library before any other \openshmem routine may
be called. \openshmem does not support communication with dynamically created
or connected processes. In such a scenario, MPI must be used to communicate

> with these processes.

Author: I am actually a little cautious in making the possible communication approaches more restricted, because users may also use other approaches such as socket. We only want to prohibit OpenSHMEM in such scenarios. How about we say?
```In such a scenario, MPI can be used to communicate with these processes.```

communication resource for \openshmem and MPI communication may maintain an
internal reference counter in order to ensure that the shared resource is
initialized only once and thus no shared resource is released until the last
finalization call is made.
}

\subsection{Dynamic Process Creation}
\label{subsec:interoperability:mpmd}

MPI defines a dynamic process model that allows creation of processes after
an MPI application has started (e.g., by calling \FUNC{MPI\_Comm\_spawn}) and
connection to independent processes (e.g., through \FUNC{MPI\_Comm\_accept}
and \FUNC{MPI\_Comm\_connect})
and provides a mechanism to establish communication
between the newly created processes and the existing MPI application (see
MPI standard version 3.1, Chapter 10).
Unlike MPI, \openshmem starts all processes at once and requires all PEs to
collectively allocate and initialize resources (e.g., symmetric heap) used by
the \openshmem library before any other \openshmem routine may
be called. Communicating with a dynamically created process in the \openshmem
environment may result in undefined behavior.
Hence, users should not use \openshmem and MPI dynamic process models

> in the same program.

Reviewer 2: I'm not sure this is the right conclusion.  Connect/accept can be used to connect a job to a service, e.g. connect a SHMEM job to a database or I/O service.  Or to connect two different SHMEM jobs as a workflow.  These should all work fine, provided the communication is done with MPI and not SHMEM.  A better conclusion may be:
```
\openshmem does not support communication with dynamically created or connected processes. 
In such a scenario, MPI must be used communicate with these processes.
```

Author: I like the suggested conclusion. I disallowed it because I think it is hard to precisely cover all combinations of `SHMEM+MPI dynamic process` (see below). 

- case 1:
After MPI initialization and before any PEs start OpenSHMEM initialization, the program spawns(connects to) a new MPI process. It is unclear whether P1 is included in the original `TEAM_WORLD`.
```
P0                  Spawned P1
MPI_Init(..)
spawn();
shmem_init();       shmem_init();
```

- case 2:
After MPI and SHMEM initialization, the program spawns(connects to) a new process and the new process wants to communicate with PEs in the original `TEAM_WORLD`.
```
P0                  Spawned P1
MPI_Init(..)
shmem_init();
spawn();
                    shmem_init();
```

- case 3:
After MPI and SHMEM initialization, the program spawns(connects to) some new processes (e.g., P1 and P2 in the example). Can `P1` and `P2` communicate with each other through OpenSHMEM? 
```
P0                  Spawned P1      Spawned P2
MPI_Init(..)
shmem_init();
spawn();
                    shmem_init();   shmem_init();
```

I think the new conclusion clearly defines all the above cases!

\subsection{Thread Safety}
\label{subsec:interoperability:thread}
Both \openshmem and MPI define the interaction with user threads in a program
with routines that can be used for initializing and querying the thread
environment. In a hybrid program, the user may request different thread levels
at the initialization calls of \openshmem and MPI environments; however, the
returned support level provided by the \openshmem library might be different

> from that returned in an \openshmem-only program. For instance, the former

Reviewer 2: Suggest:
```
returned support level provided by the \openshmem or MPI library might be different
from that returned in an non-hybrid program. For instance, the first
```

initialization call in a hybrid program may initialize a resource with the

> user-requested thread level, but the supported level cannot be updated by the latter

Reviewer 2: ```suggestion
user-requested thread level, but the supported level cannot be updated by a subsequent
```

initialization call if the underlying software runtime of \openshmem and MPI
share the same internal communication resource.
The program should always check the \VAR{provided} thread level returned
at the corresponding initialization call or query the level of thread support
after initialization to portably ensure thread support in each communication
environment.

Both \openshmem and MPI define similar thread levels, namely, \VAR{THREAD\_SINGLE},
\VAR{THREAD\_FUNNELED}, \VAR{THREAD\_SERIALIZED}, and \VAR{THREAD\_MULTIPLE}.
When requesting threading support in a hybrid program, however,
the following additional rules are applied if the implementations of \openshmem

> and MPI share the same internal communication resource.

Reviewer 2: Is it possible to query whether synchronization requirements are combined for both MPI and SHMEM or separate (i.e. separate serialization domains)?

Author: Good idea. We may also add a query property to indicate whether the same master thread is required for the `THREAD_FUNNELED` safety. But I need time to think about a use case which can benefit from the feature. Maybe it is better to first wrap up the current proposal and propose the other properties separately. What do you think?

Author: Over the WG call, we decided to postpone the query API into a separate proposal and we will work on the thread safety properties there.

    initialization call of the other model in the same program.

    \item The \VAR{THREAD\_FUNNELED} thread level allows only the main thread to
    make communication calls. A hybrid program using the \VAR{THREAD\_FUNNELED}
    thread level in both \openshmem and MPI should ensure that the same main thread
    is used in both communication environments.

    \item The \VAR{THREAD\_SERIALIZED} thread level requires the program to ensure
    that communication calls are not made concurrently by multiple threads. If a
    hybrid program uses \VAR{THREAD\_SERIALIZED} in one communication environment
    and \VAR{THREAD\_SERIALIZED} or \VAR{THREAD\_FUNNELED} in the other one, it
    should also guarantee that the \openshmem and MPI calls are not made concurrently
    from two distinct threads.
\end{itemize}

\subsection{Mapping Process Identification Numbers}
\label{subsec:interoperability:id}

Similar to the PE identifier in \openshmem, MPI defines rank as the
identification number of a process in a communicator. Both the \openshmem PE
and the MPI rank are unique integers assigned from zero to one less than the total
number of processes. In a hybrid program, the \openshmem

> PE and the MPI rank in \VAR{MPI\_COMM\_WORLD} of a process can be equal.

Reviewer 2: ```suggestion
PE number in \LibHandleRef{SHMEM\_TEAM\_WORLD} and the MPI rank in \VAR{MPI\_COMM\_WORLD} of a process can be equal.
```

This feature, however, may be provided by only some of the \openshmem and MPI
implementations (e.g., if both environments share the same underlying process
manager) and is not portably guaranteed. A portable program should always

> use the standard functions in each model, namely, \FUNC{shmem\_my\_pe} in \openshmem

Reviewer 2: ```suggestion
use the standard functions in each model, namely, \FUNC{shmem\_team\_my\_pe} in \openshmem
```

Author: Also replaced the old API with teams-API in the example code.

Author: Also replaced the old API with teams-API in the example code.

in each communication environment and manage the mapping of identifiers in the
program when necessary.

\subsubsection*{Example}
\label{subsubsec:interoperability:id:example}
The following example demonstrates how to manage the mapping between \openshmem
PE identifier and MPI ranks in \VAR{MPI\_COMM\_WORLD} in a hybrid \openshmem
and MPI program.

\lstinputlisting[language={C}, tabsize=2,
      basicstyle=\ttfamily\footnotesize]
      {example_code/hybrid_mpi_mapping_id.c}

\subsection{RMA Programming Models}
\label{subsec:interoperability:rma}

\openshmem and MPI each define similar one-sided communication models;
however, a portable program should not assume interoperability between these
models.
For instance, \openshmem guarantees the atomicity only of concurrent \openshmem AMO operations
that operate on symmetric data with the same datatype. Access to the same symmetric
object with MPI atomic operations, such as an \FUNC{MPI\_Fetch\_and\_op}, may
result in an undefined result. Furthermore,
because most RMA programs can be written by using either \openshmem or MPI RMA,
users should choose only one of the RMA models in the same program, whenever

> possible, for performance and code simplicity.

Reviewer 2: If a user is using a library that uses MPI RMA, they may not have an either-or choice.  Phase-based access (e.g. MPI I/O and SHMEM RMA phases) using the different models should still be fine.  Suggest instead:
```
Users should avoid situations where MPI and \openshmem operations perform concurrent
accesses the same memory locations.
```

Author: Makes sense.

\subsection{Mapping Process Identification Numbers}
\label{subsec:interoperability:id}

Similar to the PE number in \openshmem, MPI defines rank as the

> identification number of a process in a communicator. Both the \openshmem PE

Author: Use \ac{PE}.

Reviewer 3: FWIW, I think we can fix all the missing `\ac{...}` macros afterward as editorial changes. (MPI itself should be an acronym, no?)

Author: MPI stands for `Message Passing Interface`. Should we define the acronym macro for it?

Author: I just found that \ac{MPI} is already defined. Replaced `MPI` with `\ac{MPI}` in the proposal.

For instance, \openshmem guarantees the atomicity only of concurrent \openshmem AMO operations
that operate on symmetric data with the same datatype. Access to the same symmetric
object with MPI atomic operations, such as an \FUNC{MPI\_Fetch\_and\_op}, may
result in an undefined result. Users should avoid situations where MPI and

> \openshmem operations perform concurrent accesses to the same memory location.

Author: Make a pass to minimize usage of `users`
